common/flatpak-dir.c

```diff
@@ -6610,12 +6610,84 @@ flatpak_dir_list_app_refs_with_runtime (FlatpakDir        *self,
       FlatpakDecomposed *app_ref = g_ptr_array_index (app_refs, i);
       /* deploy v4 guarantees runtime info */
       g_autoptr(GBytes) app_deploy_data = flatpak_dir_get_deploy_data (self, app_ref, 4, NULL, NULL);
+      const char *app_runtime;
 
-      if (app_deploy_data)
+      if (app_deploy_data == NULL)
+        continue;
+
+      app_runtime = flatpak_deploy_data_get_runtime (app_deploy_data);
+      if (g_strcmp0 (app_runtime, runtime_pref) == 0)
+        g_ptr_array_add (apps, flatpak_decomposed_ref (app_ref));
+    }
+
+  return g_steal_pointer (&apps);
+}
+
+GPtrArray *
+flatpak_dir_list_app_refs_with_runtime_extension (FlatpakDir        *self,
```

Comment:
Hmm, yeah it makes sense that this implementation is very inefficient. I wonder if there's a relatively straightforward fix along the same lines as your PR https://github.com/flatpak/flatpak/pull/4576. In the meantime I split off some commits from this PR so they can be merged while I work on this.

Reply:
I think I have the performance sorted with the current version of this branch which does in-memory caching. Here are the numbers I get with `time ./flatpak uninstall --user --all | false`:
- current version with in-memory caching: about 0.28s real time
- current `main` branch which doesn't check for extension dependencies at all: about 0.21s
- previous state of this branch that does disk reads to build the dependency graph for each extension: about 3.8s

So the numbers I'm seeing aren't as egregious as what you saw, but I have fewer flatpaks installed, and we're measuring in different ways, and maybe you're not using an SSD? In any case, I think the new version of the branch is orders of magnitude better, and agree it would make sense to apply this technique more broadly across the codebase (but I'm unlikely to be able to volunteer to do that soon).